feat: Create migration from old config to new

liordino opened the pull request.

liordino committed:
fix: adapt old settings loader to the new schema (filter schema_version)

feat: migrate commissions globals

feat: migrate locales

refactor: generalize the old fields search

chore: created functions signatures for all config migrations

feat: created wallet migration

feat: migrate operator info

feat: migrate coin atm radar

feat: migrate terms and conditions

feat: migrate commissions overrides

fix: removed the wallet_COIN_active field (don't exist anymore)

chore: moved the config-migration lib to the lib folder

feat: migrate cashout configurations

feat: migrate notifications globals

feat: export migration function

feat: migrate most of notifications scoped configs

fix: added the missing text property to the terms and conditions
migration

feat: migrate compliance triggers

feat: migrate receipt printing

feat: migrate accounts

chore: remove test code form module

refactor: change some functions naming

fix: set default trigger type to 'volume'

feat: added threshold days (default 1) to triggers

fix: removed strike from the accounts importing

refactor: cleaner code on fixed properties

feat: avoid repeated crypto/machine pairs on the commissions overrides
migrations

refactor: make renameAccountFields function internal to the account
migration function

fix: migrate all crypto scoped commission overrides

liordino commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create migration from old config to new](https://trello.com/c/HKZowKPG/164-create-migration-from-old-config-to-new)

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+          p[1]
+        ])
+    )
+  }
```
This can be replaced with `_.mapKeys(it => replacements[it])` (not tested code)

this won't be needed because of the bitgo change.

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+function migrate(config, accounts) {
+  return {
+    ...migrateConfig(config),
+    ...migrateAccounts(accounts)
```
better to just return a plain object on these methods and do:

```
config: migrateConfig(config),
accounts: migrateAccounts(accounts)
```

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+    token: 'token',
+    ZECWalletId: 'zecWalletId',
+    ZECWalletPassphrase: 'zecWalletPassphrase'
+  }
```
If it's only the casing that's changing I think it's better to rename it on the new admin, this is actually a bug since the bitgo module expects upper case coins to work.

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+  if (accounts.itbit) accountsProps.push(['itbit', accounts.itbit])
+  if (accounts.kraken) accountsProps.push(['kraken', accounts.kraken])
+  if (accounts.mailgun) accountsProps.push(['mailgun', accounts.mailgun])
+  if (accounts.twilio) accountsProps.push(['twilio', accounts.twilio])
```
Instead of doing all of this the whole method will become:

``` js
const accountArray = ['bitgo', 'bitstamp'] // plus all other accounts
return _.pick(accountArray)(accounts)
```

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+
+const GLOBAL = 'global'
+const ALL_CRYPTOS = _.values(COINS).sort()
+const ALL_MACHINES = 'all machines'
```
better to make this `const ALL_MACHINES =  'ALL_MACHINES'`

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+
+  function getAllMachines(config) {
+    return _.uniq(config.map(f => f.fieldLocator.fieldScope.machine).filter(m => m !== GLOBAL))
+  }
```
No longer being used.

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+  }
+
+  const { global, scoped } = getConfigFields(_.keys(codes), config)
+  const allMachines = getAllMachines(config).sort()
```
No longer being used.

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+function migrateCommissions(config) {
+  function areArraysEquals(arr1, arr2) {
+    return _.xor(arr1, arr2).length === 0
+  }
```
Maybe change to:
`const areArraysEquals = (arr1, arr2) => _.isEmpty(_.xor(arr1, arr2))`

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,442 @@
+
+  let allCommissionsOverrides =
+    machineAndCryptoScopedPlusCryptoScopedCommissionsOverrides
+      .concat(machineScoped.map(f => {
+        const cryptosAlreadyScopedWithThisFieldMachine = _.flatten(
+          machineAndCryptoScopedPlusCryptoScopedCommissionsOverrides
+            .filter(innerF => innerF.scope.machine.includes(f.scope.machine))
+            .map(innerF => innerF.scope.crypto)
+        )
+        
+        f.scope.crypto = _.difference(
+          ALL_CRYPTOS,
+          cryptosAlreadyScopedWithThisFieldMachine
+        )
+
+        return f
```
This is pretty hard to understand. Some things you can do to make this clearer:

1. Try to make it step by step. For instance do only the machineScoped.map and create a variable for that. Afterwards you can concat in another variable.

2. Names are pretty explanatory, but it's too long for comfort. When we're reading this we only need some context if the code shows intent. 

``` js
// machineAndCryptoScopedPlusCryptoScopedCommissionsOverrides changed to withCryptoScoped
// cryptosAlreadyScopedWithThisFieldMachine changed to coinsToRemove or toRemove
```

3. Doing it more functionally (with the help of lodash) can help breaking things into smaller chunks and make intent clearer.

We can break logic into pretty bite sized chunks that are not that hard to follow:

``` js


const getMachine = _.get('scope.machine')
const getCrypto = _.get('scope.crypto')

const filteredMachineScoped = _.map(it => {
     const filterByMachine = _.filter(_.includes(getMachine(it)))
     const flattenCoins =  _.compose(_.flatten, _.map(getCrypto))
     const diffAllCryptos = _.compose(_.diference(ALL_CRYPTOS))

     const unrepeatedCryptos = _.compose(diffAllCryptos, flattenCoins, filterByMachine)(withCryptoScoped) 

     return _.set('scope.crypto', unrepeatedCryptos)(it)
)(machineScoped)

```

Note how even flattenCoins and diffAllCryptos can be moved to the outside of the function since they don't require any context. It would make the function a 3 liner.

liordino replied:
Way better! Thanks!

liordino committed:
fix: return plain objects from functions to make the jsons more readable

fix: fix bitgo fields casing

fix: improve commissions migration function readability

refactor: standard styling

liordino committed:
feat: add fallback values to the migration

liordino committed:
feat: created db migration for the new config

liordino committed:
feat: create migration to move machine names from file to db

fix: updates machine names before the config migration

fix: load machineLoader

fix: create a param to ignore the schema version when loading the latest
config using the old loader

liordino commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create a migration to update the machine names on the db](https://trello.com/c/1DpEXWcL/191-create-a-migration-to-update-the-machine-names-on-the-db)

RafaelTaranto commented on lib/new-settings-loader.js:
```diff
@@ -1,71 +1,114 @@
-let db = null
-
-// TODO new-admin save to actual db, like the old settings
-low(adapter).then(it => {
-  db = it
-})
+const NEW_SETTINGS_LOADER_SCHEMA_VERSION = 2
```
Wasn't this already merged in another PR? There's something strange here.

liordino replied:
Yeah, it's already on dev. Dunno why it's considered a change here

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,462 @@
+    triggers.push(
+      createTrigger(requirements.sanctions, global.smsVerificationThreshold)
+    )
+  }
```
Check if active and threshold are set before creating the trigger

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -0,0 +1,462 @@
+    thresholdDays = 1,
+    triggerType = triggerTypes.volume,
+    id = uuid.v4(),
+    cashDirection = 'both'
```
On this file you don't need all of this. Just set it straight on the return statement

liordino committed:
refactor: remove unnecessary arguments on createTrigger function

fix: check if there's an smsVerificationThreshold configured prior to
migrating triggers

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -386,6 +382,11 @@ function migrateComplianceTriggers (config) {
     getConfigFields(codes, config).global.map(f => [f.code, f.value])
   )
 
+  if (!_.isNumber(global.smsVerificationThreshold))
+    return {
+      triggers: []
+    }
+
```
I don't think we want this. There can be other requirements active without a sms threshold. What I meant on the my comment was to test both *VerificationActive and *Threshold for every IF of the triggers creation logic.

liordino committed:
fix: migrate triggers with the correct thresholds and verify if they're
valid
